python/cudf/cudf/core/internals/where.py

```diff
@@ -0,0 +1,375 @@
+            warnings.warn(
+                f"Type-casting from {other_col.dtype} "
+                f"to {source_col.dtype}, there could be potential data loss"
+            )
+        return source_col, other_col.astype(source_col.dtype)
+    else:
+        common_dtype = cudf.utils.dtypes.find_common_type(
+            [source_col.dtype, other_col.dtype]
+        )
+        return source_col.astype(common_dtype), other_col.astype(common_dtype)
+
+
+def _check_and_cast_columns_with_scalar(source_col, other_scalar, inplace):
+    """
+    Returns type-casted column `source_col` & scalar `other_scalar`
+    based on `inplace` parameter.
+    """
+    if cudf.utils.dtypes.is_categorical_dtype(source_col.dtype):
+        return source_col, other_scalar
+
+    device_scalar = cudf.Scalar(
+        _normalize_scalars(source_col, other_scalar),
+        dtype=source_col.dtype if other_scalar is None else None,
+    )
+
+    if other_scalar is None:
+        return source_col, device_scalar
+    elif cudf.utils.dtypes.is_mixed_with_object_dtype(
+        device_scalar, source_col
+    ):
+        raise TypeError(
+            "cudf does not support mixed types, please type-cast "
+            "the column of dataframe/series and other "
+            "to same dtypes."
+        )
+    if inplace:
+        if not np.can_cast(device_scalar, source_col.dtype):
+            warnings.warn(
+                f"Type-casting from {device_scalar.dtype} "
+                f"to {source_col.dtype}, there could be potential data loss"
+            )
+        return source_col, device_scalar.astype(source_col.dtype)
+    else:
+        if pd.api.types.is_numeric_dtype(source_col.dtype) and np.can_cast(
+            other_scalar, source_col.dtype
+        ):
+            common_dtype = source_col.dtype
+        else:
+            common_dtype = cudf.utils.dtypes.find_common_type(
+                [source_col.dtype, np.min_scalar_type(other_scalar)]
+            )
+
+        source_col = source_col.astype(common_dtype)
+        return source_col, cudf.Scalar(other_scalar, dtype=common_dtype)
+
+
+def _normalize_columns_and_scalars_type(frame, other, inplace=False):
+    """
+    Try to normalize the other's dtypes as per frame.
+
+    Parameters
+    ----------
+
+    frame : Can be a DataFrame or Series or Index
+    other : Can be a DataFrame, Series, Index, Array
+        like object or a scalar value
+
+        if frame is DataFrame, other can be only a
+        scalar or array like with size of number of columns
+        in DataFrame or a DataFrame with same dimension
+
+        if frame is Series, other can be only a scalar or
+        a series like with same length as frame
+
+    Returns:
+    --------
+    A dataframe/series/list/scalar form of normalized other
+    """
+    if isinstance(frame, cudf.DataFrame) and isinstance(other, cudf.DataFrame):
+        source_df = frame.copy()
+        other_df = other.copy()
```

Comment:
Could we use shallow copies instead of deep copies here?

Reply:
Yep